src/java/org/apache/cassandra/index/sai/utils/RangeConcatIterator.java

```diff
@@ -0,0 +1,124 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.index.sai.utils;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.PriorityQueue;
+
+import org.apache.cassandra.io.util.FileUtils;
+
+/**
+ * {@link RangeConcatIterator} takes a list of sorted range iterator and concatenates them, leaving duplicates in
+ * place, to produce a new stably sorted iterator. Duplicates are eliminated later in
+ * {@link org.apache.cassandra.index.sai.plan.StorageAttachedIndexSearcher}
+ * as results from multiple SSTable indexes and their respective segments are consumed.
+ *
+ * ex. (1, 2, 3) + (3, 3, 4, 5) -> (1, 2, 3, 3, 3, 4, 5)
+ * ex. (1, 2, 2, 3) + (3, 4, 4, 6, 6, 7) -> (1, 2, 2, 3, 3, 4, 4, 6, 6, 7)
+ *
+ */
+public class RangeConcatIterator extends RangeIterator
+{
+    private final PriorityQueue<RangeIterator> ranges;
+    private final List<RangeIterator> toRelease;
+
+    protected RangeConcatIterator(RangeIterator.Builder.Statistics statistics, PriorityQueue<RangeIterator> ranges)
+    {
+        super(statistics);
+
+        this.ranges = ranges;
+        this.toRelease = new ArrayList<>(ranges);
+    }
+
+    @Override
+    @SuppressWarnings("resource")
+    protected void performSkipTo(PrimaryKey primaryKey)
+    {
+        while (!ranges.isEmpty())
+        {
+            if (ranges.peek().getCurrent().compareTo(primaryKey) >= 0)
+                break;
+
+            RangeIterator head = ranges.poll();
+
+            if (head.getMaximum().compareTo(primaryKey) >= 0)
+            {
+                head.skipTo(primaryKey);
+                ranges.add(head);
+                break;
+            }
+        }
+    }
+
+    @Override
+    public void close() throws IOException
+    {
+        // due to lazy key fetching, we cannot close iterator immediately
+        toRelease.forEach(FileUtils::closeQuietly);
+    }
+
+    @Override
+    @SuppressWarnings("resource")
+    protected PrimaryKey computeNext()
+    {
+        while (!ranges.isEmpty())
+        {
+            RangeIterator current = ranges.poll();
+            if (current.hasNext())
+            {
+                PrimaryKey next = current.next();
+                // hasNext will update RangeIterator's current which is used to sort in PQ
+                if (current.hasNext())
+                    ranges.add(current);
```

Comment:
Aside: I guess `testDuplicatedElementsAndOverlappingBoundaries` illustrates why we have to suffer w/ adding the head element back after consumption...and why we need the PQ in the first place for the concat.

Reply:
I took a stab at just rewriting this without the PQ. Here's what I ended up w/ and it doesn't seem to fail any of the tests:

```
private int i = 0;
...
protected void performSkipTo(PrimaryKey primaryKey)
{
    while (i < toRelease.size())
    {
        RangeIterator currentIterator = toRelease.get(i);

        if (currentIterator.getCurrent().compareTo(primaryKey) >= 0)
            break;

        if (currentIterator.getMaximum().compareTo(primaryKey) >= 0)
        {
            currentIterator.skipTo(primaryKey);
            break;
        }

        i++;
    }
}
...
protected PrimaryKey computeNext()
{
    while (i < toRelease.size())
    {
        RangeIterator currentIterator = toRelease.get(i);
        
        if (currentIterator.hasNext())
            return currentIterator.next();
        
        i++;
    }

    return endOfData();
}
```

Reply:
I'll answer a few of above comments about the use of `PriorityQueue` here. As you have noted, it is only used by the `KeyRangeConcatIterator` (new name), so it makes no sense to create one for the other builders. 

It looks to me like it would far more sense to defer more of the `Builder` and `Statistics` work to the iterator classes. 

Another example of this is the `disJoint` calculation. This is only used by intersections so it makes no sense to be calculating it for the other iterators.

In the case of the use of `PriorityQueue` in the concat iterator, I would prefer to keep it as is for the time being. This iterator is used heavily with the on-disk indexes and I am concerned about the performance impact of any change. I would prefer that we create separate follow up ticket (or just add a TODO to that class) to investigate this later.

Reply:
@mike-tr-adamson I'm fine w/ a TODO in the class that references a follow-up Jira (that can preserve the code from my comment above).

> This iterator is used heavily with the on-disk indexes and I am concerned about the performance impact of any change

I guess I'm concerned about the opposite (introducing a subtle correctness problem), although perhaps the performance win of ditching the PQ is more complicated than I'm making it.

Reply:
I have raised [CASSANDRA-18165](https://issues.apache.org/jira/browse/CASSANDRA-18165) for this and will add a TODO referencing this ticket to the `KeyRangeConcatIterator`.

Reply:
Thanks!